contracts/roulette/lib.rs

```diff
@@ -0,0 +1,388 @@
+            self.env().block_number() > self.betting_period_end()
+        }
+
+        /// Returns true if there is still place for more bets
+        #[ink(message)]
+        pub fn are_bets_accepted(&self) -> bool {
+            let data = self.data.get().unwrap();
+            data.next_bet_id < data.maximal_number_of_bets.into()
+        }
+
+        /// Place a bet
+        ///
+        /// Places a bet from a player along for the native amount of token included in the transaction
+        #[ink(message, payable)]
+        pub fn place_bet(&mut self, bet_type: BetType) -> Result<()> {
+            let player = self.env().caller();
+            let amount = self.env().transferred_value();
+
+            let mut data = self.data.get().unwrap();
+
+            let next_bet_id = data.next_bet_id;
+            if self.is_betting_period_over() || self.are_bets_accepted() {
+                return Err(RouletteError::NoMoreBetsCanBeMade);
+            };
+
+            if amount < data.minimal_bet_amount {
+                return Err(RouletteError::BetAmountIsTooSmall);
+            }
+
+            if self.account_id_to_bet.contains(player) {
+                return Err(RouletteError::PlayerAlreadyPlacedABet);
+            }
+
+            let bet = Bet {
+                player,
+                amount,
+                bet_type,
+            };
+
+            let potential_payout = calculate_payout(&bet, None);
+            let casino_balance = self.env().balance();
+
+            // casino checks if it has enough balance to cover all the winnings up to this point
+            if data
+                .potential_payouts
+                .checked_add(potential_payout)
+                .ok_or(RouletteError::ArithmethicError)?
+                > casino_balance
+            {
+                return Err(RouletteError::NotEnoughBalance);
+            }
+
+            data.next_bet_id = data
+                .next_bet_id
+                .checked_add(1)
+                .ok_or(RouletteError::ArithmethicError)?;
+
+            self.data.set(&data);
+            self.account_id_to_bet.insert(player, &next_bet_id);
+            self.bets.insert(next_bet_id, &bet);
+
+            Self::emit_event(
+                self.env(),
+                Event::BetPlaced(BetPlaced {
+                    player,
+                    amount,
+                    bet_type,
+                }),
+            );
+
+            Ok(())
+        }
+
+        /// Spin the wheel
+        ///
+        /// Will also distribute the winnings to the players and reset the state, starting a new round of bets
+        #[ink(message)]
+        pub fn spin(&mut self) -> Result<()> {
+            if !self.is_betting_period_over() {
+                return Err(RouletteError::BettingPeriodNotOver);
+            };
+
+            // generate a "random" number between 1 and 36
+            // NOTE: this is a very poor source of randomness, we should add rcf palet
```

Comment:
or use block hash

Reply:
What is rcf pallet? This would be a very good thing to spend some time teaching about. how some randomness sources seem good at a glance, but can be attacked by miners.

If you do want to add rfc pallet, it would be nice to expose it to the evm too, perhaps through a precompile. But that is just me daydreaming. A more urgent goal is getting 5 or 6 contracts like this one sketched out in both ink and solidity. my brainstorm ideas are https://github.com/Polkadot-Blockchain-Academy/Contract-Writing-Workshop/tree/main/evm/src

Reply:
it is (insecure) randomness collective flip - this pallet can be passed as randomness source to pallet contracts, but for some time it is kinda deprecated

Reply:
I guess this is a nice place to pause and discuss on-chain randomness during lectures, and we can leave it at that for now